runtime/compiler/env/JITServerPersistentCHTable.cpp

```diff
@@ -0,0 +1,626 @@
+   if (!removeStr.empty())
+      commitRemoves(removeStr);
+
+#ifdef COLLECT_CHTABLE_STATS
+   uint32_t nBytes = removeStr.size() + modifyStr.size();
+   CHTABLE_UPDATE_COUNTER(_updateBytes, nBytes);
+   CHTABLE_UPDATE_COUNTER(_numUpdates, 1);
+   uint32_t prevMax = _maxUpdateBytes;
+   _maxUpdateBytes = std::max(nBytes, _maxUpdateBytes);
+#endif
+   //if (modifyStr.size() + removeStr.size() != 0)
+      //fprintf(stderr, "CHTable updated with %d bytes\n", modifyStr.size() + removeStr.size());
+   }
+
+void 
+JITServerPersistentCHTable::commitRemoves(const std::string &rawData)
+   {
+   auto &data = getData();
+   TR_OpaqueClassBlock **ptr = (TR_OpaqueClassBlock**)&rawData[0];
+   size_t num = rawData.size() / sizeof(TR_OpaqueClassBlock*);
+   for (size_t i = 0; i < num; i++)
+      {
+      auto item = data[ptr[i]];
+      data.erase(ptr[i]);
+      if (item) // may have already been removed earlier in this update block
+         jitPersistentFree(item);
+      }
+   CHTABLE_UPDATE_COUNTER(_numClassesRemoved, num);
+   }
+
+void 
+JITServerPersistentCHTable::commitModifications(const std::string &rawData)
+   {
+   auto &data = getData();
+   std::unordered_map<TR_OpaqueClassBlock*, std::pair<FlatPersistentClassInfo*, TR_PersistentClassInfo*>> infoMap;
+
+   // First, process all TR_PersistentClassInfo entries that have been
+   // modified at the client without worrying about subclasses
+   size_t bytesRead = 0;
+   size_t count = 0;
+   while (bytesRead != rawData.length())
+      {
+      TR_ASSERT(bytesRead < rawData.length(), "Corrupt CHTable!!");
+      FlatPersistentClassInfo *info = (FlatPersistentClassInfo*)&rawData[bytesRead];
+      TR_OpaqueClassBlock *classId = (TR_OpaqueClassBlock *) (((uintptr_t) info->_classId) & ~(uintptr_t)1);
+      TR_ASSERT(classId, "Class cannot be null (on server)");
+      TR_PersistentClassInfo* clazz = findClassInfo(classId);
+      if (!clazz)
+         {
+         clazz = new (PERSISTENT_NEW) TR_PersistentClassInfo(NULL);
+         data.insert({classId, clazz});
+         }
+      infoMap.insert({classId, {info, clazz}});
+      // Overwrite existing TR_PersistentClassInfo entry with info from client
+      // Do not touch the subclasses list at this point
+      bytesRead += FlatPersistentClassInfo::deserializeClassSimple(clazz, info);
+      count++;
+      }
+
+   // populate subclasses
+   for (auto it : infoMap)
+      {
+      auto flat = it.second.first;
+      auto persist = it.second.second;
+      persist->removeSubClasses();
+      for (size_t i = 0; i < flat->_numSubClasses; i++)
+         {
+         auto classInfo = findClassInfo(flat->_subClasses[i]);
+
+         // For some reason the subclass info is still null sometimes. This may be indicative of a larger problem or it could be harmless.
+         // JITServer TODO: figure out why this is happening
+         //TR_ASSERT(classInfo, "subclass info cannot be null: ensure subclasses are loaded before superclass");
```

Comment:
I would replace this assert with a FATAL one. This is a bug that needs to be corrected.